runtime/browser/ui/desktop/exclusive_access_bubble.cc

```diff
@@ -0,0 +1,138 @@
+// Copyright (c) 2012 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "xwalk/runtime/browser/ui/desktop/exclusive_access_bubble.h"
+
+#include "base/strings/utf_string_conversions.h"
+#include "ui/base/l10n/l10n_util.h"
+#include "ui/gfx/geometry/rect.h"
+#include "ui/strings/grit/ui_strings.h"
+#include "xwalk/grit/xwalk_resources.h"
+
+// NOTE(koz): Linux doesn't use the thick shadowed border, so we add padding
+// here.
+#if defined(OS_LINUX)
+const int ExclusiveAccessBubble::kPaddingPx = 8;
+#else
+const int ExclusiveAccessBubble::kPaddingPx = 15;
+#endif
+const int ExclusiveAccessBubble::kInitialDelayMs = 3800;
+const int ExclusiveAccessBubble::kIdleTimeMs = 2300;
+const int ExclusiveAccessBubble::kSnoozeNotificationsTimeMs = 900000;  // 15m.
+const int ExclusiveAccessBubble::kPositionCheckHz = 10;
+const int ExclusiveAccessBubble::kSlideInRegionHeightPx = 4;
+const int ExclusiveAccessBubble::kSlideInDurationMs = 350;
+const int ExclusiveAccessBubble::kSlideOutDurationMs = 700;
+const int ExclusiveAccessBubble::kPopupTopPx = 45;
+
+ExclusiveAccessBubble::ExclusiveAccessBubble(const GURL& url)
+    : url_(url) {
+}
+
+ExclusiveAccessBubble::~ExclusiveAccessBubble() {
+}
+
+void ExclusiveAccessBubble::StartWatchingMouse() {
+  // Start the initial delay timer and begin watching the mouse.
+  hide_timeout_.Start(FROM_HERE,
+                      base::TimeDelta::FromMilliseconds(kInitialDelayMs),
+                      this, &ExclusiveAccessBubble::CheckMousePosition);
+
+  gfx::Point cursor_pos = GetCursorScreenPoint();
+  last_mouse_pos_ = cursor_pos;
+  mouse_position_checker_.Start(
+      FROM_HERE, base::TimeDelta::FromMilliseconds(1000 / kPositionCheckHz),
+      this, &ExclusiveAccessBubble::CheckMousePosition);
+}
+
+void ExclusiveAccessBubble::StopWatchingMouse() {
+  hide_timeout_.Stop();
+  idle_timeout_.Stop();
+  mouse_position_checker_.Stop();
+}
+
+bool ExclusiveAccessBubble::IsWatchingMouse() const {
+  return mouse_position_checker_.IsRunning();
+}
+
+void ExclusiveAccessBubble::CheckMousePosition() {
+  // Desired behavior (without the "simplified" flag):
+  //
+  // +------------+-----------------------------+------------+
+  // | _  _  _  _ | Exit full screen mode (F11) | _  _  _  _ |  Slide-in region
```

Comment:
I think we'd rather need "close" instead of "Exit full screen mode", for the fullscreen apps it should not be possible to switch to windowed mode. @darktears WDYT?